Sources/CNIOLinux/liburing_shims.c

```diff
@@ -0,0 +1,545 @@
+    return (getenv("SWIFTNIO_URING_USE_NEW_POLL") != NULL)?1:0;
+}
+
+int CNIOLinux_io_uring_load()
+{
+    void *dl_handle;
+    const char *err;
+    
+    // first a number of checks, did we compile with actual liburing headers?
+#ifdef C_NIO_LIBURING_UNAVAILABLE
+    return -1;
+#endif
+
+    // do the kernel have io_uring syscalls?
+    if (_check_syscall_available() != 0) {
+        fprintf(stderr, "io_uring syscall not available.\n");
+        return -1;
+    }
+
+    // have we manually enabled liburing?
+    if (getenv("SWIFTNIO_URING_ENABLED") == NULL)
+    {
+        return -1;
+    }
+    
+    // have we explicitly diabled liburing?
+    if (getenv("SWIFTNIO_URING_DISABLED") != NULL) // Just an esacpe hatch - allows for easier testing with epoll
+    {
+        fprintf(stderr, "SWIFTNIO_URING_DISABLED set, disabling liburing.\n");
+        return -1;
+    }
+    
+    // then we can finally try to load the library and resolve all symbols
+    dlerror(); // canonical way of clearing dlerror
+    dl_handle = dlopen("liburing.so", RTLD_LAZY);
+    if (((err = dlerror()) != NULL) || !dl_handle) {
+        fprintf(stderr, "WARNING: Failed to load liburing.so, using epoll() instead. [%s] [%p]\n", err?err:"Unknown reason", dl_handle);
+        return -1;
+    }
+    
+     // try to resolve all symbols we need, macro will fail with -1 if unsuccessful
+    _DL_RESOLVE(io_uring_get_probe_ring);
+    _DL_RESOLVE(io_uring_get_probe);
+    _DL_RESOLVE(io_uring_free_probe);
+    _DL_RESOLVE(io_uring_queue_init_params);
+    _DL_RESOLVE(io_uring_queue_init);
+    _DL_RESOLVE(io_uring_queue_mmap);
+    _DL_RESOLVE(io_uring_ring_dontfork);
+    _DL_RESOLVE(io_uring_queue_exit);
+    _DL_RESOLVE(io_uring_peek_batch_cqe);
+    _DL_RESOLVE(io_uring_wait_cqes);
+    _DL_RESOLVE(io_uring_wait_cqe_timeout);
+    _DL_RESOLVE(io_uring_submit);
+    _DL_RESOLVE(io_uring_submit_and_wait);
+    _DL_RESOLVE(io_uring_get_sqe);
+    _DL_RESOLVE(io_uring_register_buffers);
+    _DL_RESOLVE(io_uring_unregister_buffers);
+    _DL_RESOLVE(io_uring_register_files);
+    _DL_RESOLVE(io_uring_unregister_files);
+    _DL_RESOLVE(io_uring_register_files_update);
+    _DL_RESOLVE(io_uring_register_eventfd);
+    _DL_RESOLVE(io_uring_register_eventfd_async);
+    _DL_RESOLVE(io_uring_unregister_eventfd);
+    _DL_RESOLVE(io_uring_register_probe);
+    _DL_RESOLVE(io_uring_register_personality);
+    _DL_RESOLVE(io_uring_unregister_personality);
+    _DL_RESOLVE(__io_uring_sqring_wait);
+    _DL_RESOLVE(__io_uring_get_cqe);
+        
+    // Finally after resolving, probe actual capabilities
+    if (_check_capabilities() != 0) {
+        (void) dlclose(dl_handle); // we don't care about errors, nothing we can do anyway
+        return -1;
+    }
+
+    return 0;
+}
+
+// And the wrappers, should never be called unless we've done CNIOLinux_io_uring_load once first.
+
+pthread_mutex_t global_ring_mutex = PTHREAD_MUTEX_INITIALIZER;
```

Comment:
- this adds a global, non-namespaced symbol which NIO promised not to ever do outside a major revision.
- why do we need a global mutex?

Reply:
It should have been static, its just serialising the creation of the shared uring for sqpoll, if multiple event loops initialise rings simultaneously, only one of them should set up the shared ring (which is created a kernel-side polling thread that all event loops will share - basically removing syscalls from submission path during load (kernel thread is dropped if inactive for 1s, then recreated as needed))